proposals/gc/MVP-JS.md

```diff
@@ -0,0 +1,532 @@
+from a wasm imported function), the instantiation algorithm must ensure that
+imported type definitions' fields' types are compatible. For example, given the
+following module:
+
+```wat
+;; example5.wat --> example5.wasm
+(module
+    (type $Point (import "" "Point") (eq (struct (field i32) (field i32))))
+    (type $Rect (import "" "Rect") (eq (struct (field (ref $Point)) (field (ref $Point)))))
+    ;; ...
+)
+```
+
+this instantiation would succeed:
+
+```js
+const Point = new StructType([{type: int32}, {type: int32}]);
+const Rect = new StructType([{type: Point.ref}, {type: Point.ref}]);
+
+WebAssembly.instantiateStreaming(fetch('example5.wasm'), {'':{Point, Rect}});
+```
+
+while this instantiation would fail because of the `Point1`/`Point2` mismatch in the
+field types of `Rect`:
+
+```js
+const Point1 = new StructType([{type: int32}, {type: int32}]);
+const Point2 = new StructType([{type: int32}, {type: int32}]);
+const Rect = new StructType([{type: Point2.ref}, {type: Point2.ref}]);
+
+WebAssembly.instantiateStreaming(fetch('example5.wasm'), {'':{Point1, Rect}});
+```
+
+This instantiation-time checking would extend to the signatures of imported
+typed functions as an extension of the existing wasm-to-wasm import type checking
+performed by [instantiate_module](https://webassembly.github.io/spec/core/appendix/embedding.html#embed-instantiate-module).
+
+Additionally, using type imports of [Web IDL interface objects](https://heycam.github.io/webidl/#interface-object),
+the signatures of [Web IDL methods](https://heycam.github.io/webidl/#idl-operations) could
+be checked at instantiation-time, eliminating the normal dynamic type checks
+performed by JavaScript at call-time. **In this way, the preservation of nominal
+typing via type imports can yield performance benefits, not just error-catching
+benefits.**
+
+
+## Other WebAssembly Value Type objects
+
+In addition to `WebAssembly.ref(T)`, these additional Value Type object singletons
+and constructor functions would be added to the `WebAssembly` namespace to
+reflect all the other field types in the GC proposal:
+* `WebAssembly.funcref` : constructor function reflecting [`anyfunc`/`funcref`](https://github.com/WebAssembly/reference-types/blob/master/proposals/reference-types/Overview.md#language-extensions).
+* `WebAssembly.eqref` : singleton reflecting [`eqref`](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#value-types)
+  whose `[[Write]]` accepts only objects, symbols and `null`.
+* `WebAssembly.optref(T)` : constructor function reflecting [`optref <typeidx>`](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#value-types),
+  for a given type definition `T`.
+* `WebAssembly.i31ref` : singleton reflecting [`i31ref`](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#value-types)
+  whose `[[Write]]` only accepts integral `Number` and `BigInt` values in the range [-2<sup>31</sup>, 2<sup>31</sup>-1].
```

Comment:
I believe negative values could only go down to -2^30.

Reply:
D'oh!